ZOOKEEPER-3342: Use StandardCharsets

belugabehr opened the pull request:
> Encodes this String into a sequence of bytes using the platform's default charset, storing the result into a new byte array.  The behavior of this method when this string cannot be encoded in the default charset is unspecified.

https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#getBytes--

1.  Since this is a distributed system, it is always possible that different nodes have different default charsets defined.  I think it's most safe to specify it explicitly across all nodes for safety sake.  You could for example see a situation where an upgrade JVM uses a different default and during a rolling upgrade of the JVM, different nodes now have a different default.
2.  The default charset is usually "ISO-8859-1".  UTF-8 covers more of our international friends.
3. Explicitly specifying the CharSet yields slight performance gains 
4. Explicitly specifying the CharSet removes the need for try/catch blocks of UnsupportedEncodingException

https://blog.codecentric.de/en/2014/04/faster-cleaner-code-since-java-7/

BELUGABEHR commented:
@maoling I updated the description.  Thanks!

Note that I didn't change any of the 'four letter words' transformation to UTF-8 since they are clearly ASCII characters.  The ASCII Strings will produce the same byte array regardless of the encoding used (UTF/ISO-8859/ASCII)

enixon commented on zookeeper-server/src/test/java/org/apache/zookeeper/common/AtomicFileWritingIdiomTest.java:
```diff
@@ -348,12 +350,12 @@ private String getContent(File file, String encoding) throws IOException {
 
     private String getContent(File file) throws IOException {
-        return getContent(file, "ASCII");
+        return getContent(file, StandardCharsets.US_ASCII);
```
Do we use ASCII anywhere else?

BELUGABEHR replied:
For this test class, it's consistent in that it expects Ascii on line 356/358.  Ascii fits perfectly fine into UTF-8 so this test is not invalid and probably doesn't add any value to change it because the data being used in this test is not outside the ASCII range.

tamaashu commented:
What if someone uses ZooKeeper on Windows? Is your change not breaking backward compatibility?

"Windows applications use the UTF-16 implementation of Unicode" source: https://docs.microsoft.com/en-us/windows/desktop/intl/character-sets

BELUGABEHR commented:
@tamaashu Java also historically has used the same encoding as the one your presented.  Regardless, UTF-8 can capture all UTF-16 values (and then some).  Like all things Java, the character encoding works correctly across platforms.

https://softwareengineering.stackexchange.com/questions/174947/why-does-java-use-utf-16-for-internal-string-representation

I wouldn't recommend making the character encoding a configurable option.

1. There's no way currently to record the encoding used in all the various places ZK, so if the default changes between server restarts, reading a snapshot, reading a ZNode name, reading a ZNode value, etc. may break.
2.  Allowing for a configurable character encoding will explode the test metric for ZK.  Using UTF-8, which covers pretty much every language, will keep the testing in-check.

Since we're changing to UTF-8, which is most permissive, the chance of a backwards capability issue is very low.

BELUGABEHR commented:
http://utf8everywhere.org/

BELUGABEHR commented:
@anmolnar Can you please take a look at this again after CI tests complete? Thanks!

BELUGABEHR commented:
CI failed, but it seems that it was unable to download JDK 11 before it actually did much of anything.

maoling commented:
@BELUGABEHR 
git fetch upstream and push again,CI will not fail due to [PR-1005](https://github.com/apache/zookeeper/pull/1005) has been merged.

anmolnar commented:
retest maven build

eolivelli commented:
You have checkstyle violations

belugabehr commented:
Jenkins build failed on UT:

```
2019-10-07 19:57:59,516 [myid:] - WARN  [WorkerSender[myid=3]:QuorumCnxManager@661] - Cannot open channel to 5 at election address /127.0.0.1:16766
java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.zookeeper.server.quorum.QuorumCnxManager.connectOne(QuorumCnxManager.java:627)
	at org.apache.zookeeper.server.quorum.QuorumCnxManager.connectOne(QuorumCnxManager.java:687)
	at org.apache.zookeeper.server.quorum.QuorumCnxManager.toSend(QuorumCnxManager.java:599)
	at org.apache.zookeeper.server.quorum.FastLeaderElection$Messenger$WorkerSender.process(FastLeaderElection.java:494)
	at org.apache.zookeeper.server.quorum.FastLeaderElection$Messenger$WorkerSender.run(FastLeaderElection.java:478)
	at java.lang.Thread.run(Thread.java:748)
```

I do not think it's related to this PR

belugabehr commented:
@eolivelli @anmolnar Could you please take a look at this one again?

Thanks!

belugabehr commented:
@eolivelli Before I need to fix conflicts again :)

Can you please consider for merge?

belugabehr commented:
@anmolnar Good for 3.7?

anmolnar commented:
@belugabehr Sorry for ignoring this so long. Yes, I think it's good for 3.7, but you have to rebase again.

ctubbsii commented on zookeeper-recipes/zookeeper-recipes-queue/src/test/java/org/apache/zookeeper/recipes/queue/DistributedQueueTest.java:
```diff
@@ -27,243 +28,260 @@
+    } catch (NoSuchElementException e) {
+      return;
+    }
+    Assert.assertTrue(false);
```
Not sure why this was changed, but it should be changed back:

```suggestion
    Assert.fail();
```

(or even better, use assertThrows instead, but that's outside the scope of this changeset)

belugabehr replied:
@ctubbsii Thanks for the review.  I'm not sure how these changes came to be. I don't think I changed them manually and instead came from something I did merging this relatively old branch into master branch.  I am working on taking into account some of your feedback now and these types of changes were reverted when I tried to rebase on master again.  Thanks.

ctubbsii commented on zookeeper-recipes/zookeeper-recipes-queue/src/test/java/org/apache/zookeeper/recipes/queue/DistributedQueueTest.java:
```diff
@@ -27,243 +28,260 @@
-    @Test
-    public void testRemove2() throws Exception {
-        createNremoveMtest("/testRemove2", 10, 2);
+    byte[] data = null;
+    for (int i = 0; i < m; i++) {
+      data = queueHandles[1].remove();
```
I think this creates an unused variable. You could remove the assignment, as it was before, or you can do something with it in a new Assert statement. A simple thing is just assert that it's not null.

ctubbsii commented on zookeeper-recipes/zookeeper-recipes-queue/src/test/java/org/apache/zookeeper/recipes/queue/DistributedQueueTest.java:
```diff
@@ -27,243 +28,260 @@
-            queueHandles[i] = new DistributedQueue(clients[i], dir, null);
-        }
+    final byte[] takeResult[] = new byte[1][];
+    Thread takeThread = new Thread() {
+      public void run() {
```
Why convert lambda to an inner class? This seems like a regression.

ctubbsii commented on zookeeper-recipes/zookeeper-recipes-queue/src/test/java/org/apache/zookeeper/recipes/queue/DistributedQueueTest.java:
```diff
@@ -27,243 +28,260 @@
 
-        for (int i = 0; i < n; i++) {
-            String offerString = testString + i;
-            queueHandles[0].offer(offerString.getBytes());
-        }
+        } catch (InterruptedException e) {
```
Why separate out these exceptions instead of keeping the original multi-catch? This seems like a regression.

ctubbsii commented on zookeeper-recipes/zookeeper-recipes-queue/src/test/java/org/apache/zookeeper/recipes/queue/DistributedQueueTest.java:
```diff
@@ -27,243 +28,260 @@
+        } catch (KeeperException e) {
 
-    @Test
-    public void testElement2() throws Exception {
-        createNremoveMelementTest("/testElement2", 10, 2);
-    }
+        } catch (InterruptedException e) {
 
-    @Test
-    public void testElement3() throws Exception {
-        createNremoveMelementTest("/testElement3", 1000, 500);
+        }
+      }
+    };
```
Same comments about keeping lambda and multi-catch

ctubbsii commented on zookeeper-recipes/zookeeper-recipes-queue/src/test/java/org/apache/zookeeper/recipes/queue/DistributedQueueTest.java:
```diff
@@ -27,243 +28,260 @@
+
+    takeThread.join();
+
+    Assert.assertTrue(takeResult[0] != null);
```
assertNotNull is better. This should go back to the way it was.

ctubbsii commented on zookeeper-recipes/zookeeper-recipes-queue/src/test/java/org/apache/zookeeper/recipes/queue/DistributedQueueTest.java:
```diff
@@ -27,243 +28,260 @@
     }
+    int numAttempts = 2;
+    for (int i = 0; i < numAttempts; i++) {
+      final byte[] takeResult[] = new byte[1][];
```
The type of the variable is `byte[][]`, so this should go back to the way it was.

ctubbsii commented on zookeeper-recipes/zookeeper-recipes-queue/src/test/java/org/apache/zookeeper/recipes/queue/DistributedQueueTest.java:
```diff
@@ -27,243 +28,260 @@
+    for (int i = 0; i < numAttempts; i++) {
+      final byte[] takeResult[] = new byte[1][];
+      final String threadTestString = testString + i;
+      Thread takeThread = new Thread() {
+        public void run() {
+          try {
+            takeResult[0] = queueHandles[0].take();
+          } catch (KeeperException e) {
+
+          } catch (InterruptedException e) {
+
+          }
+        }
+      };
```
Another conversion of lambda/multi-catch to the old style code that should go back to the way it was. There's more below. I'll stop calling them out, but they should be changed back everywhere.

ctubbsii commented on zookeeper-recipes/zookeeper-recipes-queue/src/test/java/org/apache/zookeeper/recipes/queue/DistributedQueueTest.java:
```diff
@@ -27,243 +28,260 @@
-            Assert.assertNotNull(takeResult[0]);
-            Assert.assertEquals(new String(takeResult[0]), threadTestString);
-        }
+      Assert.assertTrue(takeResult[0] != null);
```
should go back to assertNotNull

ctubbsii commented on zookeeper-server/src/main/java/org/apache/zookeeper/ServerAdminClient.java:
```diff
@@ -52,7 +53,7 @@ public static void ruok(String host, int port) {
 
             int rc = is.read(resBytes);
-            String retv = new String(resBytes);
+            String retv = new String(resBytes, StandardCharsets.UTF_8);
```
These would be great to use static imports for the `UTF_8`, as was done in some of the other classes.

ctubbsii commented on zookeeper-server/src/main/java/org/apache/zookeeper/cli/ReconfigCommand.java:
```diff
@@ -152,7 +153,7 @@ public boolean exec() throws CliException {
 
             byte[] curConfig = ((ZooKeeperAdmin) zk).reconfigure(joining, leaving, members, version, stat);
-            out.println("Committed new configuration:\n" + new String(curConfig));
+            out.println("Committed new configuration:\n" + new String(curConfig, StandardCharsets.UTF_8));
```
This should probably just be two `println` statements instead of appending a newline and using string concatenation.

ctubbsii reviewed (changes requested):
I found several regressions in these changes that should go back to the way they were. Specifically, the use of inner classes and separate catch blocks when there used to be lambdas and multi-catch. Also, there was some changes to Assert methods use that didn't make sense, like changing `Assert.assertNotNull(x)` to `Assert.assertTrue(x != null)`, `Assert.fail()` to `Assert.assertTrue(false)`, etc. that should go back to what they were before.

Also, I think you should consistently use static imports to import the charsets from `StandardCharsets`. It makes the code much more readable, and `StandardCharsets` is one of the few classes where static imports is a good idea, because the names are obvious and don't collide with anything else. It's used in some places, which is great, but not others. I think it should be used consistently.

I did like the way you fixed some of the whitespace indentation for the changed lines. It makes the code more readable... even though looking at the diff was a little trickier. Luckily, GitHub makes it easy to ignore whitespace changes in the diff while reviewing, and so does the git command line.

Overall, I think it's a good idea to use StandardCharsets, but this needs a tiny bit of polishing.

belugabehr commented:
@ctubbsii Thanks for the review and re-sparking my attention on this PR.  I have gone through, added your suggestions, and cleaned up all the weirdness.  I look forward to your next review.

ctubbsii reviewed (approved):
Changes look good to me.

belugabehr commented:
@ctubbsii  @anmolnar Thanks for the +1's.

Unit tests seem to be failing repeatedly with transient errors, something different any time.  Possible for a merge regardless?

eolivelli commented:
Usually we don't have many flaky tests.
We must see a green build.

I will review the patch ASAP, hopefully we will be able to commit it as soon as CI passes

eolivelli reviewed (approved):
LGTM

The change may be considered a breaking change.
We will have to state it clearly into the release notes

belugabehr commented:
@eolivelli Third time is a charm.  Thanks!
